akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala

```diff
@@ -268,4 +277,44 @@ import akka.util.Timeout
    * INTERNAL API
    */
   @InternalApi private[akka] def messageAdapters: List[(Class[_], Any => T)] = _messageAdapters
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def setCurrentActorThread(): Unit = {
+    _currentActorThread match {
+      case OptionVal.None =>
+        _currentActorThread = OptionVal.Some(Thread.currentThread())
+      case OptionVal.Some(t) =>
+        throw new IllegalStateException(
+          s"Invalid access by thread from the outside of $self. " +
+          s"Current message is processed by $t, but also accessed from from ${Thread.currentThread()}.")
+    }
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def clearCurrentActorThread(): Unit = {
+    _currentActorThread = OptionVal.None
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def checkCurrentActorThread(operationName: String): Unit = {
```

Comment:
I would have preferred not having a string `operationName` here since it will be in the stack trace of the exception (or we could even pick it up from the stack trace if we really think it is important in the message as well)

Reply:
Good idea, I have removed it. Let's rely on that the stack trace is logged, which it should.